src/browser/wait.js

```diff
@@ -0,0 +1,173 @@
+		}
+
+		if (element.children) {
+			Array.from(element.children).forEach(this._addShadowRoots);
+		}
+	}
+}
+
+function getUpdatePromises(element, updateAwaiters) {
+	if (element.updateComplete) {
+		updateAwaiters.push(element.updateComplete);
+	}
+
+	if (element.shadowRoot) {
+		Array.from(element.shadowRoot.children).forEach(child => getUpdatePromises(child, updateAwaiters));
+	}
+
+	if (element.children) {
+		Array.from(element.children).forEach(child => getUpdatePromises(child, updateAwaiters));
+	}
+};
+
+/**
+ * Waits for an element (and its children) to finish updating. While there is no
+ * foolproof way to ensure that the element will no longer update without user
+ * interaction, this function performs a number of checks in a loop, and is able
+ * to handle most causes of element changes to safely wait for it to fully
+ * finish updating. Additional checks can be added to the loop by providing a
+ * customAwaiter function in the options object.
+ *
+ * @param {Element} element - The element to await
+ * @param {?Object} [options]
+ * @param {number} [options.timeout=0] - Timeout in milliseconds (0 for none)
+ * @param {boolean} [options.failOnTimeout=true] - If true (default), this
+ *     function will throw an error if the timeout is exceeded. Otherwise, it
+ *     will simply return normally after the timeout expires.
+ * @param {boolean} [options.awaitHypermedia=true} - If true (default), wait for
+ *     all hypermedia requests to finish.
+ * @param {?function(Element):Promise<boolean>} [options.customAwaiter] - An
+ *     optional additional function to await on. If the function returns a
+ *     Promise that resolves to a truthy value, all checks will be rerun and the
+ *     function will be called again until it returns a falsey value or the
+ *     timeout is reached.
+ * @returns {Element} The HTML element passed into the first parameter
+ * @throws {Error} If options.timeout is set and options.failOnTimeout is not
+ *     set to false, then an error will be thrown on a timeout.
+ */
+export function waitUntilSettled(element, options) {
+	options = Object.assign(
+		{
+			timeout: 0,
+			failOnTimeout: true,
+			awaitHypermedia: true,
+			customAwaiter: null,
+		},
+		options,
+	);
+
+	if (options.timeout <= 0) {
+		return _waitUntilSettledImpl(element, options);
+	} else if (options.failOnTimeout) {
+		return Promise.race([
+			_waitUntilSettledImpl(element, options),
+			aTimeout(options.timeout).then(() => element)
+		]);
+	} else {
+		return Promise.race([
+			_waitUntilSettledImpl(element, options),
+			aTimeout(options.timeout).then(() => {
+				throw new Error('Timeout waiting for element to finish updating');
+			})
+		]);
+	}
+};
+
+async function _waitUntilSettledImpl(element, options) {
+	const hasSirenActionQueue = options.awaitHypermedia &&
+		window.D2L &&
+		D2L.Siren &&
+		D2L.Siren.ActionQueue &&
+		D2L.Siren.ActionQueue.isPending &&
+		D2L.Siren.ActionQueue.enqueue;
```

Comment:
This feels like too much D2L-specific knowledge leakage into what we're hoping to keep as a bare bones testing library. Have you looked at the [getLoadingComplete stuff](https://github.com/BrightspaceUI/testing?tab=readme-ov-file#waiting-for-asynchronous-components) ([or mixin](https://github.com/BrightspaceUI/core/tree/main/mixins/loading-complete)) to handle this kind of thing? Each component that does async calls would need to be aware of it, but that's the way we'd always imagined this stuff working.

If you did want to try and write a single thing that "waits for things to settle", you could still use `getLoadingComplete` in a shared mixin and put this logic there.

Alternately, you could create your own `fixture` method that calls into `@brightspace-ui/testing`'s `fixture` (which calls into the `open-wc` one) but then additionally waits for more stuff?

Reply:
Hmm I'm not sure... there's a lot of Siren-specific stuff in here so maybe as a testing utility in `siren-sdk`? Or does your team have a shared utils library?

Reply:
No, but I guess we can just make one